pageserver/src/tenant/timeline.rs

```diff
@@ -4571,17 +4572,16 @@ impl<'a> TimelineWriter<'a> {
         }
     }
 
-    /// Put a batch keys at the specified Lsns.
+    /// Put a batch of keys at the specified Lsns.
     ///
-    /// The batch should be sorted by Lsn such that it's safe
-    /// to roll the open layer mid batch.
+    /// The batch is sorted by Lsn (enforced by usage of [`utils::VecMap`].
     pub(crate) async fn put_batch(
         &mut self,
-        batch: Vec<(Key, Lsn, Value)>,
+        batch: VecMap<Lsn, (Key, Value)>,
         ctx: &RequestContext,
     ) -> anyhow::Result<()> {
-        for (key, lsn, val) in batch {
-            self.put(key, lsn, &val, ctx).await?
+        for (lsn, (key, val)) in batch.as_slice() {
+            self.put(*key, *lsn, val, ctx).await?
```

Comment:
This copies the `key` and `lsn`, right? I'd like to avoid that. You could add a method that consumes the vec map and iterate on that. Something like:

```
impl VecMap {
  pub fn into_inner(self) -> Vec<...> {
    self.0
  }
}
```

Reply:
Yes that is true. I would rather implement `IntoIter` for VecMap since that should do the same